### PR TITLE
[SYCL][ESIMD] Add driver option to enable SYCL Explicit SIMD extension.

### PR DESCRIPTION
**Pull request by kbobrovs**
Options name: -fsycl-explicit-simd.
This option is temporary until ESIMD and normal SYCL kernels can
co-exist in the same source.

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Comment by keryell** [comment]
Is it explained somewhere what is ESIMD?

**Comment by erichkeane** [comment]
This patch needs to come with a LanguageExtensions.rst entry that explains what this is/does.  Please add that first.

**Comment by kbobrovs** [comment]
> Is it explained somewhere what is ESIMD?

Yes. https://github.com/intel/llvm/pull/1731

**Comment by kbobrovs** [comment]
> This patch needs to come with a LanguageExtensions.rst entry that explains what this is/does. Please add that first.

This is the spec of the extension, same documentation style as all other extensions
https://github.com/intel/llvm/pull/1731

**Comment by bader** [comment]
> Options name: -fsycl-explicit-simd.
> This option is temporary until ESIMD and normal SYCL kernels can
> co-exist in the same source.
> 

How this option help "ESIMD and normal SYCL kernels to co-exist in the same source"?
Why not just emit an error (or warning) instead of adding a new option?
NOTE: diagnostics can be controlled with existing options e.g. disabled or converted from a warning to an error.

**Comment by kbobrovs** [comment]
> > Options name: -fsycl-explicit-simd.
> > This option is temporary until ESIMD and normal SYCL kernels can
> > co-exist in the same source.
> 
> How this option helps "ESIMD and normal SYCL kernels to co-exist in the same source"?
> Why not just emit an error (or warning) instead of adding a new option?
> NOTE: diagnostics can be controlled with existing options e.g. disabled or converted from a warning to an error.

It does not help co-existence. What it does is
(1) makes compiler set the __SYCL_EXPLICIT_SIMD__ macro, which is then used in headers in few places to change behavior
(2) makes FE change behavior in some cases - e.g. to allow non-constant globals

**Comment by kbobrovs** [comment]
See more details in https://github.com/intel/llvm/pull/1731

**Comment by bader** [comment]
> > > Options name: -fsycl-explicit-simd.
> > > This option is temporary until ESIMD and normal SYCL kernels can
> > > co-exist in the same source.
> > 
> > 
> > How this option helps "ESIMD and normal SYCL kernels to co-exist in the same source"?
> > Why not just emit an error (or warning) instead of adding a new option?
> > NOTE: diagnostics can be controlled with existing options e.g. disabled or converted from a warning to an error.
> 
> It does not help co-existence. What it does is
> (1) makes compiler set the **SYCL_EXPLICIT_SIMD** macro, which is then used in headers in few places to change behavior
> (2) makes FE change behavior in some cases - e.g. to allow non-constant globals

Sorry, but I have a few more questions:

(1) Why this macro should be enabled by the compiler rather than developer or in some library header? Why this macro name does not start with the underscore, which is reserved by the C++ spec for the compiler needs? Where can I find the information about the behavior it's supposed to change?
(2) Do you have a list of FE changes requiring a new driver option? Allowing non-constant globals make sense in regular SYCL mode as well. OpenCL 2.+ environment supports it. Can we have a separate option to relax this restriction or just lift it for post-SYCL-1.2.1? Will it be enough?

> See more details in #1731

I can't find anything related to new driver option, macro or FE changed behavior in this document. Should it be updated?

Is it related somehow to [implementation restrictions](https://github.com/intel/llvm/blob/f4445e43f8914437a8e0d2503239ea1862d7b4f0/sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md#implementation-restrictions)?

**Comment by kbobrovs** [comment]
> 
> (1) Why this macro should be enabled by the compiler rather than developer or in some library header?

This is similar to SYCL_EXTERNAL - I think there was earlier discussion that it is safer to enable in the compiler. But I think it can be defined in library header - should not be a problem. But in this case user would have to be very careful about header inclusion order, and make sure defining header is always included even if it is not needed otherwise. Do you think it is better? Why?

> Why this macro name does not start with the underscore, which is reserved by the C++ spec for the compiler needs?

This is github swallowing the `__`. It is `__SYCL_EXPLICIT_SIMD__`

> Where can I find the information about the behavior it's supposed to change?

Do you mean `__SYCL_EXPLICIT_SIMD__` macro? This will be in the runtime patch. For now you can access our dpcpp-cm gitlab repo and grep the runtime sources - let me know if you need a link.

> (2) Do you have a list of FE changes requiring a new driver option? Allowing non-constant globals make sense in regular SYCL mode as well. OpenCL 2.+ environment supports it. Can we have a separate option to relax this restriction or just lift it for post-SYCL-1.2.1? Will it be enough?

LLVM opts:
- `--disable-llvm-passes` is not added
- 3 new ESIMD-specific LLVM passes added
- A number of existing LLVM passes added after FE and before SPIRV generation
FE:
- Alignment attribute on return value is not added to improve code gen. @DenisBakhvalov may comment more on this one
- Non-const globals are allowed
- Zero initialization for non-const globals is skipped
- (Not implemented yet) Some of the semantics checks identifying unsupported cases. This is TBD.

But, as I said in #1731, eventually we'd like to remove that option. So can we go a slightly different way here - start with `-fsycl-explicit-simd`, which isolates the rest of SYCL from ESIMD-related modification., then incrementally eliminate the differences between ESIMD and normal SYCL. Once they are all eliminated, we safely get rid of (deprecate) the `-fsycl-explicit-simd`.
?

> 
> > See more details in #1731
> 
> I can't find anything related to new driver option, macro or FE changed behavior in this document. Should it be updated?

I was actually referring to the PR description, not to the doc. The doc is a spec. I think we should have a different doc describing ESIMD compiler usage and temporary restrictions.

> 
> Is it related somehow to [implementation restrictions](https://github.com/intel/llvm/blob/f4445e43f8914437a8e0d2503239ea1862d7b4f0/sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md#implementation-restrictions)?

Indirectly - via the `__SYCL_EXPLICIT_SIMD__` macro changing runtime behavior in some cases. Directly - enforcing extra semantics checks in the FE (see notes above)

**Comment by kbobrovs** [comment]
@bader, please comment

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -6145,6 +6149,12 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       // doing the host pass.
       CmdArgs.push_back("-fsycl");
       CmdArgs.push_back("-fsycl-is-host");
+
+      // TODO add support to specify options shared between host and device
```
Comment (erichkeane) [inline clang/lib/Driver/ToolChains/Clang.cpp:6153]: As this TODO doesn't apply to this patch in particular, can this be removed from this patch/hoisted into a better more general location? @mdtoguchi

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:6153]: I recall a different patch where I requested a note so we could be sure to consolidate.  I'm OK with removing from this patch.

Comment (kbobrovs) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Mike, I filed https://github.com/intel/llvm/issues/1808 not to loose this note.

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Awesome, thanks @kbobrovs

**Pull request merged**